### PR TITLE
feat: add secp256r1 support

### PR DESCRIPTION
**Pull request by DreamWuGit**
### Description

This PR aims to support secp256r1(p256) precompile from circuit side, below are works on signature circuit side, not include evm circuit(buss mapping), l2geth_util etc.
- [x] refactor struct `SignData` to generic type `SignData<Fq, Affine>`
- [x] add p256 event to `PrecompileEvents` and provide method to get p256 signature data
- [x] halo2-lib: update secp256r1 field(Fp, Fq) implement ScalarField.
- [x] assign new  p256 signature data into sig table
- [x] sig circuit: refactor `assign_ecdsa` to support new secp256r1
- [x] sig circuit: refactor `sign_data_decomposition` to support secp256r1
- [x] sig circuit: refactor `assign_sig_verify` to support secp256r1 
- [x] halo2-lib: update `is_on_curve_or_infinity` applicable to secp256r1
- [ ] skip keccak lookup for p256 ?
- [x] adjust params `MAX_NUM_SIG`, `LOOKUP_CELLS_PER_SIG`, `CELLS_PER_SIG` ?
- [ ] testing
  - [x]  p256 signature circuit unit test
  - [ ] ci (evm/super circuit) test
  - [ ] integration test

### Issue Link

[_link issue here_]

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

**Comment by lispc** [comment]
another test needs to be pay attention to:
   vk should be fixed, for 2 blocks, 1 has 1 r1 2 k1, the other has 2 r1 1 k1.

**Comment by lispc** [comment]
i think it is ok to first merge sig circuit changes, then second PR includes evm / bus mapping changes.

**Comment by DreamWuGit** [comment]
here only testing sig circuit CI, and will open another PR for evm/bus mapping changes later.

**Comment by lispc** [comment]
seems the TyposCheck is quite useful

File: eth-types/src/sign_types.rs [hunk 1]
```diff
@@ -139,6 +231,18 @@ impl Default for SignData {
     }
 }
 
+// Default for secp256r1
+impl Default for SignData<Fq_R1, Secp256r1Affine> {
+    // Hardcoded valid signature corresponding to a hardcoded private key and
+    // message hash generated from "nothing up my sleeve" values to make the
```
Comment (lispc) [inline eth-types/src/sign_types.rs:242]: message hash generated from "nothing up my sleeve": is this valid?  where is the hashing of this example data ""nothing up my sleeve""?

File: zkevm-circuits/src/sig_circuit.rs [hunk 2]
```diff
@@ -123,20 +133,53 @@ impl<F: Field> SubCircuitConfig<F> for SigCircuitConfig<F> {
...
         //
         // TODO: make those parameters tunable from a config file
-        let ecdsa_config = FpConfig::configure(
+        let limb_bits = 88;
+        let num_limbs = 3;
+
+        let range = RangeConfig::<F>::configure(
             meta,
-            FpStrategy::Simple,
+            RangeStrategy::Vertical,
```
Comment (lispc) [inline zkevm-circuits/src/sig_circuit.rs:140]: why this change?

Comment (lispc) [reply zkevm-circuits/src/sig_circuit.rs:140]: ok seems this is the inner struct? Range chip

File: zkevm-circuits/src/sig_circuit.rs [hunk 3]
```diff
@@ -268,12 +331,14 @@ impl<F: Field> SubCircuit<F> for SigCircuit<F> {
...
-            + block.precompile_events.get_ecrecover_events().len();
+            + block.precompile_events.get_ecrecover_events().len()
+            + block.precompile_events.get_p256_verify_events().len();
+
         // Reserve one ecdsa verification for padding tx such that the bad case in which some tx
         // calls MAX_NUM_SIG - 1 ecrecover precompile won't happen. If that case happens, the sig
         // circuit won't have more space for the padding tx's ECDSA verification. Then the
         // prover won't be able to produce any valid proof.
-        let max_num_verif = MAX_NUM_SIG - 1;
+        let max_num_verif = MAX_NUM_SIG_K1 + MAX_NUM_SIG_R1 - 1;
```
Comment (lispc) [inline zkevm-circuits/src/sig_circuit.rs]: this is wrong.  i explained the result is not (num_k1 + num_r1) / (limit_k1 + limi_r1), but max(num_k1/limit_k1, num_r1/limi_r1)

Comment (lispc) [reply zkevm-circuits/src/sig_circuit.rs]: btw i don't know why the "-1" is here. you can read the codes for why there is a "-1".

File: zkevm-circuits/src/sig_circuit.rs [hunk 4]
```diff
@@ -446,6 +513,202 @@ impl<F: Field> SigCircuit<F> {
...
+                ctx,
+                &pk_assigned,
+                &integer_r,
+                &integer_s,
+                &msg_hash,
+                4,
+                4,
+            );
+
+        println!("sig_is_valid {:?}", sig_is_valid);
```
Comment (lispc) [inline zkevm-circuits/src/sig_circuit.rs]: many println here.  can be cleaned up

File: zkevm-circuits/src/sig_circuit.rs [hunk 5]
```diff
@@ -446,6 +513,202 @@ impl<F: Field> SigCircuit<F> {
...
+        );
+
+        println!("assigned_y_tmp {:?}", assigned_y_tmp);
+        //let ecc_chip = EccChip::<F, FpChipK1<F>>::construct(ecdsa_chip.clone());
+
+        let ecc_chip = EccChip::construct(ecdsa_chip.clone());
+        ecc_chip
+            .field_chip
+            .range
+            .range_check(ctx, &assigned_y_tmp, 87);
```
Comment (lispc) [inline zkevm-circuits/src/sig_circuit.rs]: a lot of 87/88 in this file.  Make a constant?

File: zkevm-circuits/src/sig_circuit.rs [hunk 6]
```diff
@@ -644,15 +912,20 @@ impl<F: Field> SigCircuit<F> {
         })
     }
 
+    // this helper support both secp256k1 snf secp256r1
```
Comment (lispc) [inline zkevm-circuits/src/sig_circuit.rs]: `snf` what is this?

File: zkevm-circuits/src/sig_circuit.rs [hunk 7]
```diff
@@ -836,9 +1152,39 @@ impl<F: Field> SigCircuit<F> {
                     .into_iter()
                     .unzip();
 
+                let (assigned_keccak_values_r1, assigned_sig_values_r1): (
```
Comment (lispc) [inline zkevm-circuits/src/sig_circuit.rs:1133]: does r1 needs keccak? `assigned_keccak_values_r1`

File: zkevm-circuits/src/sig_circuit/utils.rs [hunk 8]
```diff
@@ -7,12 +7,16 @@ use halo2_ecc::{
...
+    halo2curves::secp256k1::{Fp as Fp_K1, Fq as Fq_K1},
+    halo2curves::secp256r1::{Fp as Fp_R1, Fq as Fq_R1},
 };
 
 // Hard coded parameters.
 // FIXME: allow for a configurable param.
-pub(super) const MAX_NUM_SIG: usize = 128;
+pub(super) const MAX_NUM_SIG_K1: usize = 128;
+// MAX_NUM_SIG/4 for MAX_NUM_SIG_R1 ?
+pub(super) const MAX_NUM_SIG_R1: usize = 32;
```
Comment (lispc) [inline zkevm-circuits/src/sig_circuit/utils.rs]: change these 2 values to 110 and 15.

File: zkevm-circuits/src/witness/block.rs [hunk 9]
```diff
@@ -167,6 +174,15 @@ impl Block {
         signatures
     }
 
+    /// Get signature (witness) from the block's p256Veirfy calls.
+    pub(crate) fn get_sign_data_p256(
+        &self,
+        padding: bool,
+    ) -> Vec<SignData<secp256r1::Fq, Secp256r1Affine>> {
+        // TODO: handle padding ?
```
Comment (lispc) [inline zkevm-circuits/src/witness/block.rs]: remember to make a decision here.

**Comment by lispc** [comment]
general seems good. 

1.  need to fix ccc 
2. there are many commented codes and println.  need to clean
3. there are some TODOs, need to make final deicisions on them

File: eth-types/src/sign_types.rs (hunk 1 above)
Comment (DreamWuGit) [reply eth-types/src/sign_types.rs:242]: it is for padding sig circuit rows, the message hash is not from real world tx but valid for signature verification. in other words. just found  pair value ( msg_hash, private key etc)  and get a valid signature.      for k1 default padding data, refer to https://github.com/scroll-tech/zkevm-circuits/blob/feat/secp256r1/eth-types/src/sign_types.rs#L223, which use dummy tx with a specified private key and sign to get a valid signature.
for p256 precompile, no need raw bytes, just get msg_hash from input data, and then do verify.

File: zkevm-circuits/src/sig_circuit/utils.rs (hunk 8 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/sig_circuit/utils.rs]: updated in [78d85e8](https://github.com/scroll-tech/zkevm-circuits/pull/1351/commits/78d85e82537accb6d0352043a08f2029e1b0c771)

File: zkevm-circuits/src/sig_circuit.rs (hunk 4 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/sig_circuit.rs]: removed

File: zkevm-circuits/src/sig_circuit.rs (hunk 2 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/sig_circuit.rs:140]: yes, take this range chip out of FpChip so that two FpChips can share only one range chip

File: zkevm-circuits/src/sig_circuit.rs (hunk 3 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/sig_circuit.rs]: will change to max_num_verif =  max(num_k1/limit_k1, num_r1/limi_r1) - 2  as last tx may include both ecrecover and p256verify at the same time. need to preserve two.

File: zkevm-circuits/src/sig_circuit.rs (hunk 5 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/sig_circuit.rs]: good point , will update

**Comment by DreamWuGit** [comment]
> general seems good.
> 
> 1. need to fix ccc
> 2. there are many commented codes and println.  need to clean
> 3. there are some TODOs, need to make final deicisions on them

done !

**Comment by lispc** [comment]
@roynalnaruto would you have some time to do a quick review on this?
